### PR TITLE
Add optical absorption model with imported data

### PR DESCRIPTION
**Pull request by hhollenb**
The base optical `Model` class with `AbsorptionModel` as an example case. `ImportedModels` is a list of `ImportOpticalModel`s with a map for built-in model classes to their imported identifier, and can be built directly or from imported data with a `ModelImporter` (included in a later PR). `ImportedModelAdapter` can then be used by models to access their MFP tables.

MFP tables are built with `detail::MfpBuilder` which internally handles the grid IDs that a model builds. It should probably be refactored when we refactor the rest of the grid builders.

For testing I created a `MockImportData` class to handle a bunch of common mock data and functions that the model classes use. I find that it makes writing the tests easy and fairly consistent, but could probably use a good critique.

### CODE REVIEW

File: src/celeritas/optical/Model.hh
Comment (sethrj) [inline src/celeritas/optical/Model.hh]:
```suggestion
class Model : public OpticalStepActionInterface,
```
in `celeritas/optical/action/ActionInterface.hh`. I think we could move all the models to the `action` directory too?

File: src/celeritas/optical/Model.hh
Comment (sethrj) [inline src/celeritas/optical/Model.hh]: This is unnecessary since the base class is already a virtual destructor.

File: src/celeritas/optical/model/AbsorptionModel.cc
Comment (sethrj) [inline src/celeritas/optical/model/AbsorptionModel.cc]: I think we'll have a lot of implementations look like this: perhaps we *should* have `build_mfp(OpticalMaterialId, MfpBuilder&)`? Or analogously to the ORANGE proto builder, have the `MfpBuilder` store the current material ID being built?

File: src/celeritas/optical/detail/MfpBuilder.hh
Comment (sethrj) [inline src/celeritas/optical/detail/MfpBuilder.hh]: Since all volumetric optical models (internal or external) will need to use this class to initialize, it shouldn't be in the detail namespace. I don't think we'll want to refactor into the generic grid inserter; instead this class may "have a" grid inserter.

File: src/celeritas/optical/model/AbsorptionModel.hh
Comment (sethrj) [inline src/celeritas/optical/model/AbsorptionModel.hh]: `override final` is redundant:
```suggestion
    void build_mfps(detail::MfpBuilder&) const final;

    // Execute the model with host data
    void step(CoreParams const&, CoreStateHost&) const final;

    // Execute the model with device data
    void step(CoreParams const&, CoreStateDevice&) const final;
```

Also marking the class as `final` will improve error checking (and, potentially, allow some small optimizations to take place).

File: src/celeritas/optical/ImportedModelAdapter.hh
Comment (sethrj) [inline src/celeritas/optical/ImportedModelAdapter.hh]: I don't think it's necessary to have builtin model classes since it's not possible for users to extend the enumerations. For now let's assume that the only models are the ones we've "hardcoded" (i.e., `range(IMC::size_)`)  into the import data, and worry later about extensibility.

File: src/celeritas/optical/ImportedModelAdapter.cc
Comment (sethrj) [inline src/celeritas/optical/ImportedModelAdapter.cc]: Instead of the separate insert and find, you should be able to do this by `insert`ing and checking the `iterator, bool` result I think. But again, perhaps for now we should just try to simplify this to include only a one-to-one mapping for the builtin models?

**Review by sethrj — COMMENTED** [review]
Sorry it took me a couple of days to review this with ongoing travel!

File: src/celeritas/optical/model/AbsorptionModel.cc
Comment (hhollenb) [reply src/celeritas/optical/model/AbsorptionModel.cc]: I think `build_mfp(OpticalMaterialId, MfpBuilder&)` is nicer because it makes `MfpBuilder` less stateful and less coupled to optical models. Looping in the `build_mfps` was a byproduct of not having an immediate way to get the number of optical materials in `optical::PhysicsParams::build_mfps`, but that should be easily resolved.

File: src/celeritas/optical/model/AbsorptionModel.hh
Comment (hhollenb) [reply src/celeritas/optical/model/AbsorptionModel.hh]: I didn't realize `final` fails for non-virtual functions that aren't overrides, that's good to know.

File: src/celeritas/optical/detail/MfpBuilder.hh
Comment (hhollenb) [reply src/celeritas/optical/detail/MfpBuilder.hh]: The refactoring was in reference to discussions about changing how core physics does their `GridBuilder`s. This is really just the same pattern that we do when building core XS tables:
 * Build data vector in reals
 * Put item ranges of built data into `GridRecord`
 * Build the grid in `...Data::grids`
 * After building grids for each material, put the range of the grid IDs into some table of type `ItemRange<GridId>`

The only difference is that in core physics the builder is cached and evaluated later, while in optical physics the grids are built when the `Model::build_mfps` function is called. There's nothing really specific to optical physics or MFPs in this code, so it feels like there should be some more common way to do this with core physics.

File: src/celeritas/optical/ImportedModelAdapter.cc
Comment (hhollenb) [reply src/celeritas/optical/ImportedModelAdapter.cc]: If we're limiting to just built-ins, then an `EnumArray` should probably work fine.

**Comment by amandalund** [comment]
Sorry for the delay @hhollenb --this looks great at first glance, I'll go over it in detail first thing tomorrow!

File: src/celeritas/optical/detail/MfpBuilder.hh
Comment (sethrj) [reply src/celeritas/optical/detail/MfpBuilder.hh]: Definitely, but as I've emphasized, the core physics cross section calculation is some of our earliest attempts to move generic data onto device, and so we don't necessarily want to emulate it: this is our chance to prototype a better way of doing it (based on some of the iterations we've made in the materials, geometry, optical properties, ...) without the baggage of needing to refactor the other physics simultaneously.

For now, let's keep it as-is so long as we get something working, and then *before* adding further models we'll consider refactors/improvements.

File: src/celeritas/optical/ImportedModelAdapter.hh
Comment (amandalund) [inline src/celeritas/optical/ImportedModelAdapter.hh:77]: Instead of constructing this with the `ImportModelId`, which is somewhat arbitrary, would it make sense to do what we do with the EM processes and construct it with the `ImportModelClass` instead? Then the models would be constructed with a shared pointer to `ImportedModels` and build the model adapter with their `ImportModelClass`.

File: src/celeritas/optical/ImportedModelAdapter.cc
Comment (amandalund) [inline src/celeritas/optical/ImportedModelAdapter.cc]: Maybe we want these to be `VALIDATE`s since we're checking input data? And we might also want to check the first vector is non-empty.

File: test/celeritas/optical/MockImportedData.hh
Comment (amandalund) [inline test/celeritas/optical/MockImportedData.hh]: If we change how the `ImportedModelAdapter` and models are constructed, we might also not need to hardcode these `ImportModelId`s in the tests.

**Review by amandalund — COMMENTED** [review]
Really nice work @hhollenb!

File: src/celeritas/optical/ImportedModelAdapter.cc
Comment (hhollenb) [reply src/celeritas/optical/ImportedModelAdapter.cc]: I believe `models_` will have to be non-empty since the check is in a loop over `models_` (the first check will be redundant).

File: src/celeritas/optical/model/AbsorptionModel.cc
Comment (amandalund) [inline src/celeritas/optical/model/AbsorptionModel.cc]: I was thinking we could construct these similar to our EM `Process`s:
```c++
using SPConstImported = std::shared_ptr<ImportedModels const>;
...
AbsorptionModel(ActionId, SPConstImported);
```
```suggestion
AbsorptionModel::AbsorptionModel(ActionId id, SPConstImported model_data)
    : Model(id, "absorption", "interact by optical absorption")
    , imported_(ImportModelClass::absorption, model_data)
```
unless there's a need to have the `ImportedModelAdapter` outside of the model.

File: src/celeritas/optical/ImportedModelAdapter.cc
Comment (amandalund) [inline src/celeritas/optical/ImportedModelAdapter.cc]:
```suggestion
    : imported_(imported)
```

File: src/celeritas/optical/ImportedModelAdapter.cc
Comment (amandalund) [reply src/celeritas/optical/ImportedModelAdapter.cc]: Ah sorry, I meant the first MFP physics vector.

File: src/celeritas/optical/model/AbsorptionModel.cc
Comment (hhollenb) [reply src/celeritas/optical/model/AbsorptionModel.cc]: Sorry, forgot to propagate the change to the actual `AbsorptionModel`.

File: src/celeritas/optical/ImportedModelAdapter.cc
Comment (hhollenb) [reply src/celeritas/optical/ImportedModelAdapter.cc]: Ah that makes sense. I should probably also check that all the MFP vectors in the table are valid as well then.

**Review by amandalund — APPROVED** [review]
Sweet, thanks @hhollenb! I'll let @sethrj take a final look before merging.

**Comment by hhollenb** [comment]
Ack sorry I clicked merge before I read you comment. I'll setup the Rayleigh scattering PR which will be mostly the same and can be used for some further review.

**Comment by sethrj** [comment]
I didn't even realize you had permissions to merge 🤔

**Pull request merged**
